File: Sources/AutoReplyEmit.swift
import Foundation

// Emits a synthesised `tm-agent reply` when the GUI-side detector observes
// the 5-line header but the agent forgot to invoke the shell command.
// Mirrors `daemon/term-meshd/src/auto_reply_emit.rs` but calls TeamDataStore
// directly (no socket round-trip) since we're already in-process.

enum AutoReplyEmit {
    private static let resultTruncChars = 1500

    /// Statuses a reply must never reopen.
    static let terminalStatuses: Set<String> = [
        "completed", "failed", "abandoned", "cancelled",
    ]

    /// Which task a reply is answering.
    ///
    /// When the caller knows — it wrote the capsule, so it does — the answer is
    /// the task the capsule named and there is nothing to decide.
    ///
    /// When it does not, this is a guess, and the shape of the guess matters.
    /// It used to be "first non-terminal in list order", with `blocked` counted
    /// as non-terminal. An unrelated turn was measured walking up to a task
    /// another turn had parked as blocked, marking it completed, and leaving
    /// its own task untouched — losing the block reason on the way.
    ///
    /// So a blocked task is now only closable by a reply that names it. Parking
    /// something is a decision; a drive-by reply is not entitled to undo one.
    /// Among the rest the most recent wins, because a reply is far likelier to
    /// answer the instruction just given than the oldest one still open.
    static func selectTask(
        from tasks: [TeamOrchestrator.TeamTask],
        preferredTaskId: String?
    ) -> TeamOrchestrator.TeamTask? {
        if let preferredTaskId,
           let named = tasks.first(where: { $0.id == preferredTaskId }) {
            return named
        }
        if let running = tasks.filter({ $0.status == "in_progress" })
            .max(by: { $0.createdAt < $1.createdAt }) {
            return running
        }
        return tasks
            .filter { !terminalStatuses.contains($0.status) && $0.status != "blocked" }
            .max(by: { $0.createdAt < $1.createdAt })
    }

    /// Normalize FULL_REPORT field: strip whitespace, treat "n/a" as nil.
    private static func normalizedFullReportPath(_ raw: String?) -> String? {
        guard let s = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !s.isEmpty else { return nil }
        if s.lowercased() == "n/a" { return nil }
        return s
    }

    /// Returns `true` when a task was matched and updated, `false` when the
    /// report file was written but no matching task existed (still useful).
    @discardableResult
    static func emit(
        teamName: String,
        agentName: String,
        event: AutoReplyEvent,
        // The task this reply is answering, when the caller knows it.
        //
        // The scrollback path never can: it sees an answer appear on a screen
        // with nothing tying it to a request. So it guesses — first
        // `in_progress`, else the first non-terminal task — and `blocked` is
        // not in that terminal set, so an unrelated turn can walk up and mark
        // a blocked task completed. Observed doing exactly that.
        //
        // A caller that sent the instruction does know, and passing it here is
        // the difference between closing a task and closing *this* task.
        preferredTaskId: String? = nil,
        agentInstanceId: String? = nil,
        store: TeamDataStore = .shared
    ) -> Bool {
        let replyText = formatReplyText(event)
        let resultPath = normalizedFullReportPath(event.fullReport)

        // 1. Find target task for this assignee (mirror Rust CLI auto-complete).
        // The assignment carries the pane identity captured at dispatch, so a
        // reply from a duplicate role can never file against its sibling.
        let tasks = store.listTasks(
            teamName: teamName,
            status: nil,
            assignee: agentName,
            needsAttention: false,
            priority: nil,
            staleOnly: false,
            dependsOn: nil
        )
        guard let task = selectTask(from: tasks, preferredTaskId: preferredTaskId),
              store.agentIdentityMatches(
                  teamName: teamName, agentName: agentName,
                  expectedInstanceId: task.assigneeInstanceId,
                  reportedInstanceId: agentInstanceId)
        else {
            return false
        }

        guard store.writeResult(teamName: teamName, agentName: agentName,
                                agentInstanceId: agentInstanceId, taskId: task.id,
                                content: replyText, resultPath: resultPath)
        else { return false }
        store.postMessage(teamName: teamName, from: agentName, content: replyText, type: "report")

        // 2. Map STATUS to task status + drive updateTask
        let taskStatus: String
        switch event.status {
        case "BLOCKED": taskStatus = "blocked"
        case "NEEDS_REVIEW": taskStatus = "review_ready"
        default: taskStatus = "completed"
        }
        let summary = String(replyText.prefix(resultTruncChars))
        let blockedReason = (event.status == "BLOCKED" && !event.body.isEmpty) ? event.body : nil
        let reviewSummary = (event.status == "NEEDS_REVIEW" && !event.body.isEmpty) ? event.body : nil

        let prevStatus = task.status
        guard let updated = store.updateTask(
            teamName: teamName,
            taskId: task.id,
            status: taskStatus,
            result: summary,
            resultPath: resultPath,
            assignee: nil,
            blockedReason: blockedReason,
            reviewSummary: reviewSummary,
            progressNote: nil
        ) else {
            return true
        }

        // 4. Mirror teamDataTaskUpdate hooks so tm-agent wait subscribers are
        //    woken and auto-recycle counters increment — bypassed when calling
        //    store.updateTask directly without going through the socket RPC path.
        if taskStatus != prevStatus {
            _ = TermMeshDaemon.shared.rpcCallRaw(method: "events.publish", params: [
                "kind": "task_status",
                "team": teamName,
                "agent": updated.assignee ?? agentName,
                "task_id": updated.id,
                "status": taskStatus,
                "prev_status": prevStatus
            ] as [String: Any])
            if taskStatus == "completed" {
                let tn = teamName, an = updated.assignee ?? agentName, ai = updated.assigneeInstanceId
                Task { @MainActor in
                    TeamOrchestrator.shared.handleTaskCompletionForAutoRecycle(
                        teamName: tn, agentName: an, agentInstanceId: ai
                    )
                }
            }
        }
        return true
    }

    static func formatReplyText(_ event: AutoReplyEvent) -> String {
        let header = "STATUS: \(event.status)\nFILES: \(event.files)\nVERIFY: \(event.verify)\nNEXT: \(event.next)\nFULL_REPORT: \(event.fullReport)"
        return event.body.isEmpty ? header : "\(header)\n\n\(event.body)"
    }
}

File: daemon/term-meshd/src/auto_reply_emit.rs
//! Bridge from [`crate::auto_reply::AutoReplyEvent`] to the Swift app's team
//! RPCs. When the detector observes a 5-line header in agent terminal output
//! but the agent never invoked `tm-agent reply` itself, we synthesise the
//! equivalent here so the leader's `tm-agent wait` resolves.
//!
//! Sequence (mirrors what `tm-agent reply` does in Rust CLI):
//! 1. `team.report` — writes the result file + posts a message
//! 2. `team.task.list assignee=<agent>` — pick in_progress, else first non-terminal
//! 3. `team.task.update` — set status (completed | blocked | review_ready) + result/reason
//!
//! Errors are logged but never propagate up — auto-reply is a best-effort
//! safety net; failure must not break the agent's PTY reader loop.

use std::time::Duration;

use serde_json::{json, Value};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::UnixStream;
use tokio::time::timeout;

use crate::auto_reply::AutoReplyEvent;

const RPC_TIMEOUT: Duration = Duration::from_secs(5);
const RESULT_TRUNC_CHARS: usize = 1500;

/// Send the synthesised reply to the Swift app socket. Returns `Ok(true)`
/// when the task was matched and updated, `Ok(false)` when no matching task
/// existed (still a partial success — report was written), `Err` on RPC failure.
pub async fn emit(
    socket_path: &str,
    team_name: &str,
    agent_name: &str,
    event: &AutoReplyEvent,
) -> Result<bool, String> {
    let reply_text = format_reply_text(event);

    // 1. team.report (writes file + posts message)
    let mut report_params = json!({
        "team_name": team_name,
        "agent_name": agent_name,
        "content": reply_text,
    });
    if let Some(rp) = normalize_full_report_path(&event.full_report) {
        report_params["result_path"] = json!(rp);
    }
    rpc_call(socket_path, "team.report", report_params)
        .await
        .map_err(|e| format!("team.report failed: {e}"))?;

    // 2. team.task.list — find target task for this assignee
    let tasks_resp = rpc_call(
        socket_path,
        "team.task.list",
        json!({
            "team_name": team_name,
            "assignee": agent_name,
        }),
    )
    .await
    .map_err(|e| format!("team.task.list failed: {e}"))?;

    let tasks = tasks_resp
        .get("tasks")
        .and_then(|t| t.as_array())
        .cloned()
        .unwrap_or_default();

    let target = tasks
        .iter()
        .find(|t| t.get("status").and_then(|s| s.as_str()) == Some("in_progress"))
        .or_else(|| {
            tasks.iter().find(|t| {
                let st = t.get("status").and_then(|s| s.as_str()).unwrap_or("");
                !matches!(st, "completed" | "failed" | "abandoned" | "cancelled")
            })
        });

    let Some(task) = target else {
        return Ok(false);
    };
    let Some(task_id) = task.get("id").and_then(|i| i.as_str()) else {
        return Ok(false);
    };

    // 3. team.task.update with status derived from STATUS line
    let task_status = match event.status.as_str() {
        "BLOCKED" => "blocked",
        "NEEDS_REVIEW" => "review_ready",
        _ => "completed",
    };
    let summary: String = reply_text.chars().take(RESULT_TRUNC_CHARS).collect();
    let mut update_params = json!({
        "team_name": team_name,
        "task_id": task_id,
        "status": task_status,
        "result": summary,
    });
    if event.status == "BLOCKED" && !event.body.is_empty() {
        update_params["blocked_reason"] = json!(event.body.clone());
    } else if event.status == "NEEDS_REVIEW" && !event.body.is_empty() {
        update_params["review_summary"] = json!(event.body.clone());
    }
    if let Some(rp) = normalize_full_report_path(&event.full_report) {
        update_params["result_path"] = json!(rp);
    }
    rpc_call(socket_path, "team.task.update", update_params)
        .await
        .map_err(|e| format!("team.task.update failed: {e}"))?;

    Ok(true)
}

fn format_reply_text(event: &AutoReplyEvent) -> String {
    let body = if event.body.is_empty() {
        String::new()
    } else {
        format!("\n\n{}", event.body)
    };
    format!(
        "STATUS: {}\nFILES: {}\nVERIFY: {}\nNEXT: {}\nFULL_REPORT: {}{body}",
        event.status, event.files, event.verify, event.next, event.full_report
    )
}

/// Minimal JSON-RPC v2 call over Unix socket. Mirrors the pattern in
/// `http::rpc_team_socket` but lives here to keep auto-reply self-contained.
async fn rpc_call(socket_path: &str, method: &str, params: Value) -> Result<Value, String> {
    let mut stream = UnixStream::connect(socket_path)
        .await
        .map_err(|e| format!("connect: {e}"))?;
    let request = json!({
        "jsonrpc": "2.0",
        "id": 1,
        "method": method,
        "params": params,
    });
    let payload = format!("{request}\n");
    stream
        .write_all(payload.as_bytes())
        .await
        .map_err(|e| format!("write: {e}"))?;

    let mut reader = BufReader::new(stream);
    let mut attempts = 0u8;
    let response_line = loop {
        let mut line = String::new();
        let bytes = timeout(RPC_TIMEOUT, reader.read_line(&mut line))
            .await
            .map_err(|_| "read timeout".to_string())?
            .map_err(|e| format!("read: {e}"))?;
        if bytes == 0 {
            return Err("closed without response".into());
        }
        let trimmed = line.trim();
        if trimmed.is_empty() {
            attempts += 1;
            if attempts >= 8 {
                return Err("only empty lines".into());
            }
            continue;
        }
        if !trimmed.starts_with('{') {
            return Err(trimmed.to_string());
        }
        break trimmed.to_string();
    };
    let response: Value = serde_json::from_str(&response_line)
        .map_err(|e| format!("parse: {e}; raw={response_line}"))?;
    if let Some(err) = response.get("error") {
        return Err(err
            .get("message")
            .and_then(|v| v.as_str())
            .unwrap_or("rpc error")
            .to_string());
    }
    Ok(response.get("result").cloned().unwrap_or_else(|| json!({})))
}

/// Normalize FULL_REPORT field value: strip whitespace, return None for empty/"n/a".
fn normalize_full_report_path(s: &str) -> Option<&str> {
    let t = s.trim();
    if t.is_empty() || t.eq_ignore_ascii_case("n/a") {
        None
    } else {
        Some(t)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ev(status: &str, body: &str) -> AutoReplyEvent {
        AutoReplyEvent {
            status: status.to_string(),
            files: "src/foo.rs".to_string(),
            verify: "cargo test".to_string(),
            next: "NONE".to_string(),
            full_report: "n/a".to_string(),
            body: body.to_string(),
            raw: format!(
                "STATUS: {status}\nFILES: src/foo.rs\nVERIFY: cargo test\nNEXT: NONE\nFULL_REPORT: n/a\n\n{body}"
            ),
        }
    }

    #[test]
    fn format_reply_includes_header_and_body() {
        let text = format_reply_text(&ev("DONE", "shipped"));
        assert!(text.starts_with("STATUS: DONE\n"));
        assert!(text.contains("FULL_REPORT: n/a"));
        assert!(text.ends_with("shipped"));
    }

    #[test]
    fn format_reply_omits_body_separator_when_empty() {
        let text = format_reply_text(&ev("DONE", ""));
        assert!(text.ends_with("FULL_REPORT: n/a"));
        assert!(!text.contains("\n\n"));
    }
}
